llpc/lower/llpcSpirvLowerRayTracing.cpp

```diff
@@ -109,66 +109,82 @@ static unsigned TraceParamsTySize[] = {
+
+  Value *accelStruct = inst->getAccelStruct();
+  Value *rayFlags = inst->getRayFlags();
+  Value *cullMask = inst->getInstanceInclusionMask();
+  Value *sbtOffset = inst->getRayContributionToHitGroupIndex();
+  Value *sbtStride = inst->getMultiplierForGeometryContribution();
+  Value *missIndex = inst->getMissShaderIndex();
+  Value *rayOrigin = inst->getOrigin();
+  Value *rayTMin = inst->getTMin();
+  Value *rayDir = inst->getDirection();
+  Value *rayTMax = inst->getTMax();
+  Value *payload = inst->getPayload();
+  Value *paq = inst->getPaq();
+
+  m_builder->SetInsertPoint(inst);
+
+  SmallVector<Value *, 12> implCallArgs({accelStruct, rayFlags, cullMask, sbtOffset, sbtStride, missIndex, rayOrigin,
+                                         rayTMin, rayDir, rayTMax, payload, paq});
+
+  auto newCall = m_builder->CreateNamedCall(mangledName, inst->getFunctionType()->getReturnType(), implCallArgs,
+                                            {Attribute::NoUnwind, Attribute::AlwaysInline});
+
+  inst->replaceAllUsesWith(newCall);
+
+  auto func = m_module->getFunction(mangledName);
 
-  if (opcode == OpTraceRayKHR) {
+  if (func->isDeclaration()) {
+    func->setLinkage(GlobalVariable::InternalLinkage);
     func->addFnAttr(Attribute::AlwaysInline);
+
+    auto rayTracingContext = static_cast<RayTracingContext *>(m_context->getPipelineContext());
+    bool indirect = rayTracingContext->getIndirectStageMask() & ShaderStageComputeBit;
+
     auto entryBlock = BasicBlock::Create(*m_context, ".entry", func);
     m_builder->SetInsertPoint(entryBlock);
+
     auto payloadTy = rayTracingContext->getPayloadType(m_builder);
-    Value *payload = m_builder->CreateAlloca(payloadTy, SPIRAS_Private);
+    Value *localPayload = m_builder->CreateAlloca(payloadTy, SPIRAS_Private);
 
-    // When constructing the function call, we put a dummy "payload type" argument after all other arguments to indicate
-    // the actual payload size.
-    assert(func->arg_size() == TraceRayParam::TraceRayCount + 1);
-    // Copy payload variable to the global payload variable
     auto payloadArg = func->getArg(TraceRayParam::Payload);
-    auto payloadTypeArg = func->arg_end() - 1;
-    unsigned payloadArgSize = alignTo(m_module->getDataLayout().getTypeAllocSize(payloadTypeArg->getType()), 4);
+    auto paqArray = func->getArg(TraceRayParam::Paq);
+    auto payloadArgSize = m_builder->CreateExtractValue(paqArray, 0);
     const Align align = Align(4);
-    // type conversion
-    m_builder->CreateMemCpy(payload, align, payloadArg, align, payloadArgSize);
+    m_builder->CreateMemCpy(localPayload, align, payloadArg, align, payloadArgSize);
 
     SmallVector<Value *, 8> args;
-    args.push_back(m_builder->CreateLoad(payloadTy, payload));
+    args.push_back(m_builder->CreateLoad(payloadTy, localPayload));
+
+    // For trace ray entry, we decided to use <2 x i32> to pass  acceleration structureso that we can easily retrieve
```

Comment:
Typo: structureso => structure so
Please also remove the double whitespace before "acceleration"